### ISSUE TITLE
Highlights get lost all the time: please make exporting and document saving 'live'

### ISSUE THREAD

**Issue by LucenaWoods** [koreader/koreader#3911]
* KOReader version: koreader-beta-20180410
* Device: Kindle PW 2
Summary: 
Problem: All pdfs eventually crash and highlights get lost. Suggestion: Make highlight exportation and document saving "live" (that is, exported & saved with every instance of highlighting and every page turn). 
--
Hi, huge fan of your work here. I use koreader to read pdf books for about 20 hours a week (a graduate student in the humanities); consider me an extremely dedicated test dummy. 

I will mention some small but very troublesome problems regarding highlights and make a suggestion on how to fix them. 

This problem is not specific to this koreader version, it's been there for a long while. 

Problem: **All PDF files eventually crash and recent highlights get lost.** Trust me, you might test koreader for 30 minutes on a pdf and have no problems, 1 hour and it might still be ok, but in the second or third hour of reading and highlighting, eventually *all* pdfs crash. And not particularly weird pdfs; simple looking, common text-based humanities books with regular page numbers (200-400) and file sizes (3-10 mb). They all crash eventually. But this is not the main problem for me. I can live and flourish even with crashes if I'm not losing my highlights all the time. 

The main problem is that the highlights get lost and there is no convenient way to stop this. And it hurts so much. Here is what happens: 

You read a pdf book, with frequent highlights. The pdf eventually crashes (again, you should trust me on this one), either freezing completely, forcing you to restart the device or taking you back to kindle's default screen. Even if you have obsessively and compulsively exported every highlight you've made to an html file while reading, when you go back after the crash, you'll see that in the pdf file your highlights have been lost and you have gone back to the place you were in the beginning of that reading session (that is because even when "save file automatically" is turned on, koreader has not had a chance to "save" the file, which happens only when you switch to another document or exit koreader normally). Yes, the highlights you have exported manually safely sits in the html file for now, but you don't have them visible in the document you're reading, and when you keep reading from where you left off and keep highlighting and then make a second export to html, it will simply omit the previously lost highlights (because they are not there in the book, they've been lost after the crash). This really takes away a huge lot from the software's highlighting functionality.

Secondly, exporting to txt file simply doesn't work. Doesn't work with any file. It just lists the file names in your device, and does not include your highlights. 

Suggestion: **Please make highlight exporting and document saving "live".** 

That is, whenever I highlight a text, I should know that a) it is exported immediately and permanently to an html file or a txt file in the device, and b) the file/document I'm reading is saved automatically, so that when the pdf crashes (and it *will* crash), I will have lost almost nothing. I should not have to make several taps to export highlights, and to switch between documents to save the document. I should just keep reading and highlighting and whenever the crash comes, I'll just be able to pick up from where I left and feel safe and secure about all the highlights I've made. 

Again, thanks for this amazing reader, you guys are the best. 
Cheers

**Comment by Eduardomb22** [comment]
You can do this https://github.com/koreader/koreader/issues/2152#issuecomment-232265720
It will auto save your pdf when you go the next page.

**Comment by Frenzie** [comment]
>  Trust me, you might test koreader for 30 minutes on a pdf and have no problems, 1 hour and it might still be ok, but in the second or third hour of reading and highlighting, eventually all pdfs crash. And not particularly weird pdfs; simple looking, common text-based humanities books with regular page numbers (200-400) and file sizes (3-10 mb).

That's my main use case for KOReader and I've never had it crash, but perhaps I don't do enough of this highlighting stuff. Could you please save and upload the crash.log the next time this happens so we'd hopefully get a clue as to what might be going wrong?

The big issue here is the way the highlights are currently saved. I have a solution already implemented (used in the dictionary/wikipedia search history) but no safe upgrade path for regular bookmarks and highlights.

In a nutshell, saving the document info currently saves *everything*. So doing it every page is possible but not necessarily advisable compared to every few pages. We should probably set the default to a conservative value (like 20 or 30) that isn't nil for now as a workaround. I should tag a version as "stable" soon and something as potentially as dangerous as this shouldn't go in it. Would you agree with setting it to something like 20 or 30 @poire-z ?

**Comment by LucenaWoods** [comment]
Thank you for letting me know the workaround! I've set it to 0 and will see what happens. If it works it'll significantly improve my life :)

"We should probably set the default to a conservative value (like 20 or 30)" -> that is 20 or 30 pages, I presume. In my experience 20 pages is definitely going to have one or two crashes. Maybe I should mention that two unusual things about my usage is frequent highlights and _very_ frequent dictionary look ups. I have several dictionary files, some with quite large sizes (100 mb). 

If you ever get around to adding that as a menu option, I would appreciate it if you could give us an option to manually set the number of pages for every save, or for every highlight export. If not, at least keeping the workaround (the defaults.lua tweak) possible would be great.

Thanks for all the work you're putting in to improve this reader. Cheers.

**Comment by TnS-hun** [comment]
Instead of saving at every DAUTO_SAVE_PAGING_COUNT pages wouldn't it be possible to force saving at the next page change if the bookmarks or highlights have changed?

**Comment by Frenzie** [comment]
Sure, go ahead. But while writing it please keep in mind it should be intended only as a placeholder for switching over to saving immediately.

**Comment by poire-z** [comment]
Corrupted/empty settings on crash/freeze should hopefully no more happen - and losing settings and annotations should now be limited to losing 15mn of annotations & page progress (that one can reduce to 5mn via the menu), thanks to #5678 and #5687 (January 2020).
